subsys/net/ip/net_context.c

```diff
@@ -1253,14 +1263,6 @@ NET_CONN_CB(tcp_established)
 		net_pkt_unref(pkt);
 	}
 
-	/* Increment the ack */
-	context->tcp->send_ack += data_len;
-	if (tcp_flags & NET_TCP_FIN) {
-		context->tcp->send_ack += 1;
-	}
-
-	send_ack(context, &conn->remote_addr, false);
-
```

Comment:
@pfalcon "ret = packet_received(conn, pkt, context->tcp->recv_user_data);" Sending ack doesn't depend on returned value. Irrespective of return value, ACK will be prepared and send.

Reply:
> Irrespective of return value, ACK will be prepared and sent.

Yes, an ACK packet will be sent in either case. But *contents* of the ACK packet are expected to depend on app's receive callback (i.e. receive callback may affect data put into an ACK reply), specifically, the receive window: 19ff9636930f1ceabae3461a43c6c1c78cf30253 .

> I noticed an issue related to this while trying with RPL BR sample. ...

Thanks for the clear description of the problem you saw. I would say, this description should go into the commit message.

But then, we found yet another issue with callback-based data processing. And it can be classified as a case of https://cwe.mitre.org/data/definitions/405.html - for a small request, "http_prepare_and_send() create new packets until all the payload filled in buffers", and all that happens in a not very appropriate place - synchronous callback from the network stack, which affects the stack performance.

I'm afraid, we'd need to find another solution, and such solution exists - BSD Sockets. How it works is that there's decoupling between packet reception of packets by network stack and consumption by an application - they both run in separate threads and communicate via queue. There's also the flow control implemented, mandated by TCP, using the concept of "receive window": when a packet is just received from peer peer and put into the queue, the recv_wnd is decreased (and communicated back in an ACK packet), signaling the peer that it has decreased quota to send more data (which eventually may get to 0). recv_wnd is increased back only when data is actually consumed from the queue.

Due to the fact that we have usecases of both updating and not update recv_wnd, the (optional) update happens in recv callback. Thus, this patch would break this processing, and will break cases of communication large amounts of data (megabytes) over sockets.

My suggestion is to separate this patch into a separate PR, and further analyze and test it from different sides, but opinion is that we'd need to find an alternative way to address this issue. Thanks.

Reply:
> But then, we found yet another issue with callback-based data processing. And it can be classified as a case of https://cwe.mitre.org/data/definitions/405.html - for a small request, "http_prepare_and_send() create new packets until all the payload filled in buffers", and all that happens in a not very appropriate place - synchronous callback from the network stack, which affects the stack performance.

I am experimenting that in my WIP PR of RPL BR. Running a separate thread to reply websocket messages. Checking for proper solution.
https://github.com/zephyrproject-rtos/zephyr/pull/5035/commits/b61d0a5d5f3d2dd3c954cf579828725b2391860e#diff-780b2211d7789425a2584af4afc52b77R1881

Reply:
Overall, let's take it away patch from this PR, time being I will keep it in https://github.com/zephyrproject-rtos/zephyr/pull/5035. I will create a separate PR for this and list issues over there.